File: docker/api/src/utils/phone.py
"""Canonical phone-number normalization — the ONE source of truth.

Canonical form
--------------
E.164 with a leading ``+``, **PRESERVING the country code**.

The ONLY implicit default is: a **bare 10-digit** number (no ``+``, no country
code) is treated as US NANP and gets ``+1`` prepended. We NEVER strip a ``+`` and
NEVER force ``+1`` onto a number that already carries a country code. International
numbers (``+44…``, ``+52…``) are ingested and passed through UNCHANGED.

This value becomes a routing AND billing key (it is what the ``number_routing``
view / on-net resolver, the CDR ``customer_id`` join, and every DID unique index
match on), so correctness and cross-layer consistency are load-bearing.

Cross-layer contract (MUST STAY IDENTICAL)
------------------------------------------
Three implementations of THIS EXACT algorithm run on the platform and must agree
character-for-character on every input:
  * Python  — this module (``docker/api/src/utils/phone.py``)  ← REFERENCE
  * Lua     — FreeSWITCH  (``docker/freeswitch/scripts/lib/number_utils.lua``)
  * TS      — React UI    (``docker/ui/src/utils/phone.ts``)
If you change the rule here, change it in all three and re-verify the shared test
vectors below. Any divergence silently mis-routes or mis-bills calls.

Shared test vectors (all layers must produce these)
---------------------------------------------------
    normalize_e164('5551234567')      == '+15551234567'   # bare 10-digit US default
    normalize_e164('15551234567')     == '+15551234567'   # 1 + NANP
    normalize_e164('+15551234567')    == '+15551234567'   # already canonical
    normalize_e164('(555) 123-4567')  == '+15551234567'   # punctuation stripped
    normalize_e164('+44 7911 123456') == '+447911123456'  # INTERNATIONAL preserved (NOT +1)
    normalize_e164('+5215512345678')  == '+5215512345678' # MX preserved
    normalize_e164('447911123456')    -> ValueError        # no +, not NANP -> ambiguous
    normalize_e164('1001')            -> ValueError        # too short for a DID
    normalize_e164('')                -> ValueError

    normalize_forward_destination('1001')           == '1001'          # PBX extension, verbatim
    normalize_forward_destination('5551234567')     == '+15551234567'  # else full E.164
    normalize_forward_destination('+44 7911 123456')== '+447911123456'

    is_valid_e164('+447911123456') is True
    is_valid_e164('447911123456')  is False
"""

import re

__all__ = ["normalize_e164", "normalize_forward_destination", "is_valid_e164"]

# Cosmetic separators that carry no routing meaning and are dropped when
# collecting the significant digits: '+', space, '-', '(', ')', '.'. We do NOT
# strip arbitrary letters — a stray letter leaves a non-digit in the input and the
# number fails validation (fail-closed). Kept as documentation of the exact set the
# Lua/TS ports must also treat as separators; the code below uses ``str.isdigit``.
_SEPARATORS = frozenset("+ -().")

# A 3-6 digit local PBX extension (e.g. 1001) — valid ONLY as a forward/failover
# destination, never as an ownable DID.
_EXTENSION_RE = re.compile(r"\d{3,6}")


def _digits(s: str) -> str:
    """Return only the ASCII decimal digits of ``s`` (drop +, space, -, (, ), .).

    Any non-separator, non-digit character (e.g. a letter) is left in place so the
    caller's length/shape checks fail closed rather than silently accepting junk.
    """
    return "".join(ch for ch in s if ch.isdigit())


def normalize_e164(raw: str) -> str:
    """Normalize ``raw`` to canonical E.164 (``+`` + country code + subscriber).

    Rules (in order):
      1. Strip surrounding whitespace; empty -> ValueError.
      2. Remember whether the input started with ``+`` (explicit country code).
      3. Collect the digit characters (dropping + space - ( ) .).
      4. If it started with ``+``: it already carries a country code — PRESERVE it.
         Require 8-15 total digits with a non-zero leading digit; return ``+`` + digits.
      5. Bare ``1``+10 NANP (11 digits, ``1`` then ``[2-9]``): return ``+`` + digits.
      6. Bare 10-digit NANP (``[2-9]`` first): return ``+1`` + digits — the ONLY
         place ``+1`` is ever synthesized.
      7. Anything else is ambiguous -> ValueError (caller must supply full E.164).

    Args:
        raw: A phone number in any accepted human/DB format.

    Returns:
        The canonical ``+E.164`` string.

    Raises:
        ValueError: If ``raw`` is empty or cannot be unambiguously canonicalized.
    """
    if raw is None:
        raise ValueError("phone number is required")

    s = raw.strip()
    if not s:
        raise ValueError("phone number is required")

    has_plus = s.startswith("+")
    d = _digits(s)

    if has_plus:
        # Explicit country code present. PRESERVE it exactly — never re-add 1, never
        # strip. E.164 allows up to 15 digits; require a sane minimum and a non-zero
        # leading digit (country codes never start with 0).
        if 8 <= len(d) <= 15 and d[0] in "123456789":
            return "+" + d
        raise ValueError(
            f"malformed E.164 number: '{raw}' "
            "(expected '+' followed by 8-15 digits, leading digit 1-9)"
        )

    # No leading '+': only unambiguous North American Numbering Plan shapes are
    # accepted, so we never mislabel a foreign national number as US.
    if len(d) == 11 and d[0] == "1" and d[1] in "23456789":
        # 1 + 10-digit NANP (leading '1' country code typed without the '+').
        return "+" + d
    if len(d) == 10 and d[0] in "23456789":
        # Bare 10-digit US NANP — the ONLY case where we synthesize the +1.
        return "+1" + d

    raise ValueError(
        f"cannot normalize '{raw}': provide full E.164 with country code, "
        "e.g. +447911123456"
    )


def normalize_forward_destination(raw: str) -> str:
    """Normalize an RCF ``forward_to`` / ``failover_to`` destination.

    A destination may be a **local PBX extension** (3-6 bare digits, e.g. ``1001``
    for FreeSWITCH/Zoiper testing), which is kept verbatim. Anything else must be a
    full phone number and is run through :func:`normalize_e164` (so 10-digit US and
    international ``+CC`` inputs canonicalize identically to a DID).

    Args:
        raw: A forward/failover destination in any accepted format.

    Returns:
        The extension string as-is, or the canonical ``+E.164`` number.

    Raises:
        ValueError: If ``raw`` is neither a 3-6 digit extension nor a valid E.164.
    """
    if raw is None:
        raise ValueError("destination is required")

    s = raw.strip()
    # Local PBX extension: 3-6 digits, exactly. Kept verbatim (not a PSTN number).
    if _EXTENSION_RE.fullmatch(s):
        return s
    return normalize_e164(s)


def is_valid_e164(raw: str) -> bool:
    """Non-raising predicate: True iff ``raw`` normalizes to canonical E.164.

    Note this returns False for a bare 3-6 digit extension — extensions are only
    meaningful for :func:`normalize_forward_destination`, not as standalone DIDs.
    """
    try:
        normalize_e164(raw)
        return True
    except (ValueError, TypeError):
        return False

File: docker/ui/app/src/utils/phone.ts
/**
 * Phone-number normalization — the CLIENT half of app-wide E.164 canonicalization.
 *
 * ⚠️  KEEP THIS ALGORITHM IDENTICAL to the backend + telephony layers:
 *       - Python:  docker/api/src/utils/phone.py            (`normalize_number_input`)
 *       - Lua:     docker/freeswitch/scripts/lib/number_utils.lua (`normalize_number_input`)
 *     All three implement the SAME canonical spec. If you change one, change all three.
 *
 * ─── Canonical form ────────────────────────────────────────────────────────────
 * Canonical = E.164 with a leading `+`, PRESERVING the country code.
 *   - Only a BARE 10-digit number (no `+`) defaults to `+1` (US).
 *   - NEVER strip a `+` / force `+1` on a value that already carries a country code —
 *     international numbers (+44…, +52…) pass through unchanged.
 *
 * The client is PERMISSIVE: it strips formatting separators and applies the canonical
 * form, but NEVER hard-blocks a plausible value. The API is the final arbiter — it
 * returns a helpful 422 when a value is genuinely unroutable. We optimize for "the
 * happy path is clean" without ever rejecting a number the user could legitimately want.
 *
 * ─── Algorithm (normalizeNumberInput) ─────────────────────────────────────────
 *   1. s = raw.trim(); if '' → return ''
 *   2. hasPlus = s starts with '+'
 *   3. d = digits of s (strip '+', space, '-', '(', ')', '.')
 *   4. if hasPlus → return (8..15 digits ? '+'+d : s)   # preserve CC; pass malformed through
 *   5. if d.length === 11 && d[0] === '1' && d[1] ∈ 2-9        → '+'+d
 *   6. if d.length === 10 && d[0] ∈ 2-9                        → '+1'+d   # bare US default
 *   7. if /^\d{3,6}$/.test(d)                                  → d        # local extension
 *   8. otherwise                                              → s.trim()  # unknown — pass raw
 *
 * ─── Test vectors (must agree across all three layers) ─────────────────────────
 *   '5551234567'       → '+15551234567'
 *   '15551234567'      → '+15551234567'
 *   '+15551234567'     → '+15551234567'
 *   '(555) 123-4567'   → '+15551234567'
 *   '+44 7911 123456'  → '+447911123456'   (country code PRESERVED — not forced to +1)
 *   '1001'             → '1001'             (3-6 digit local extension passes through)
 */

/** Characters treated as formatting separators and stripped when collecting digits. */
const SEPARATORS_RE = /[\s\-().+]/g;

/** A valid NANP leading digit for area code / exchange is 2–9 (never 0 or 1). */
function isNanpLead(ch: string): boolean {
  return ch >= '2' && ch <= '9';
}

/**
 * Normalize a user-entered phone number to the canonical E.164 form.
 *
 * Permissive by design: unknown/short/international-but-odd inputs are returned as
 * the trimmed original rather than rejected, so the API can 422 with guidance.
 *
 * @param raw The raw string exactly as typed (may include spaces, dashes, parens).
 * @returns The canonical value, or a passthrough of the trimmed input.
 */
export function normalizeNumberInput(raw: string): string {
  const s = raw.trim();
  if (s === '') return '';

  const hasPlus = s.startsWith('+');
  const digits = s.replace(SEPARATORS_RE, '');

  // Already carries a country code (leading '+'): preserve it. Only re-emit a clean
  // '+<digits>' when the length is a plausible E.164 (8..15 digits, per ITU-T E.164).
  // If it looks malformed, pass the raw value through and let the API decide.
  if (hasPlus) {
    return digits.length >= 8 && digits.length <= 15 ? `+${digits}` : s;
  }

  // 11 digits with a US country-code prefix: 1 + valid NANP area code.
  if (digits.length === 11 && digits[0] === '1' && isNanpLead(digits[1])) {
    return `+${digits}`;
  }

  // Bare 10-digit US number (no country code): default to +1.
  if (digits.length === 10 && isNanpLead(digits[0])) {
    return `+1${digits}`;
  }

  // Local extension (3–6 digits, no separators): pass through unchanged. This is a
  // legitimate forward_to target for on-net local delivery.
  if (/^\d{3,6}$/.test(digits)) {
    return digits;
  }

  // Unknown shape — do NOT block. Return the trimmed original; the API is the arbiter.
  return s;
}

/**
 * Advisory check: does the input look like something the platform can route?
 *
 * ADVISORY ONLY. Use this for a soft hint or to enable a Save button — NEVER to
 * hard-disable submit on a plausible value. A `true` result means the normalized
 * value is either a `+`-prefixed 8..15-digit E.164 number or a 3–6 digit extension.
 *
 * @param raw The raw string exactly as typed.
 * @returns true if the normalized form looks enterable; false otherwise.
 */
export function isEnterableNumber(raw: string): boolean {
  const normalized = normalizeNumberInput(raw);
  if (normalized === '') return false;

  // Full E.164: '+' followed by 8..15 digits.
  if (/^\+\d{8,15}$/.test(normalized)) return true;

  // Local extension: 3–6 digits.
  if (/^\d{3,6}$/.test(normalized)) return true;

  return false;
}

// Re-export the display formatter so callers can import both the normalizer (for
// writes) and the formatter (for reads) from a single phone module if they prefer.
export { fmt } from './format';
